pkgs/by-name/te/teamspeak3/package.nix

```diff
@@ -0,0 +1,143 @@
+{
+  lib,
+  stdenv,
+  fetchurl,
+  fetchzip,
+  makeWrapper,
+  makeDesktopItem,
+  glib,
+  qt5,
+  libsForQt5,
+  xkeyboard_config,
+  perl,
+  libcxx,
+  autoPatchelfHook,
+}:
+
+let
+  arch = "amd64";
+
+  pluginsdk = fetchzip {
+    url = "https://files.teamspeak-services.com/releases/sdk/3.3.1/ts_sdk_3.3.1.zip";
+    hash = "sha256-wx4pBZHpFPoNvEe4xYE80KnXGVda9XcX35ho4R8QxrQ=";
+  };
+in
+
+stdenv.mkDerivation rec {
+  pname = "teamspeak-client";
+
+  version = "3.6.2";
+
+  src = fetchurl {
+    url = "https://files.teamspeak-services.com/releases/client/${version}/TeamSpeak3-Client-linux_${arch}-${version}.run";
+    hash = "sha256-WfEQQ4lxoj+QSnAOfdCoEc+Z1Oa5dbo6pFli1DsAZCI=";
+  };
+  # It's an install script
+  dontUnpack = true;
+
+  nativeBuildInputs = [
+    perl # Installer script needs `shasum`
+    qt5.wrapQtAppsHook
+    autoPatchelfHook
+  ];
+
+  buildInputs =
+    [
+      libsForQt5.quazip
+      glib
+      libcxx
+    ]
+    ++ (with qt5; [
+      qtbase
+      qtwebengine
+      qtwebchannel
+      qtwebsockets
+      qtsvg
+    ]);
+
+  # This runs the installer script. If it gets stuck, run it with -x.
+  # If it then gets stuck at something like:
+  #
+  # ++ exec
+  # + PAGER_PATH=
+  #
+  # it's looking for a dependency and didn't find it. Check the script and make
+  # sure the dep is in nativeBuildInputs.
+  buildPhase = ''
+    runHook preBuild
+
+    # Run the installer script non-interactively
+    echo -e '\ny' | PAGER=cat sh -e $src
+
+    # Delete unecessary libraries - these are provided by nixos.
+    pushd TeamSpeak3-Client-linux_amd64/
```

Comment:
I'm not really a huge fan of this. Doing this in unpackPhase means it's somewhat expected to have gone into the subdirectory that this generates at the end anyway (and probably should be specified as sourceRoot, for instance), so the `mv * ..` isn't required.

Reply:
The thing is that this needs to happen after the installer has run. It seems more sensible to me to have the installer be ran in the `buildPhase` rather than the `unpackPhase` as that's the primary process that produces the desired binaries.

I now have the directory switching encapsulated within the buildPhase though which I also like better.

Reply:
> It seems more sensible to me to have the installer be ran in the `buildPhase` rather than the `unpackPhase` as that's the primary process that produces the desired binaries.

I'd argue that we are not producing binaries, but merely extracting them. The same way we let `dpkg` extract .deb files in unpackPhase, I'd argree with @lukegb, that we should extract these files durin unpackPhase. That would also allow people to add `patches` and other tasks to the patch phases.